src/x3f_process.c

```diff
@@ -66,40 +60,111 @@ static int get_black_level(x3f_t *x3f,
+		   "Right" /* Only used in printout */
+  };
+  int use[4] = {1, 1, 1, 1};
+  x3f_area16_t area[4];
+
+  if (image->channels < colors) return 0;
+
+  /* Workaround for bug in DP2 firmware. DarkShieldBottom is specified
+     incorrectly and thus ingnored. */
+#define BOTTOM 1
+  use[BOTTOM] = (!x3f_get_prop_entry(x3f, "CAMMODEL", &cammodel) ||
+		 strcmp(cammodel, "SIGMA DP2"));
+
+  /* Real CAMF rects */
+  for (i=0; i<2; i++)
+    if (use[i])
+      use[i] = x3f_crop_area_camf(x3f, name[i], image, rescale, &area[i]);
+
+  /* Column based rects */
+  for (i=2; i<4; i++)
+    if (use[i])
+      use[i] = x3f_crop_area_column(x3f, side[i], image, rescale, &area[i]);
+
+  for (i=0; i<4; i++)
+    if (use[i])
+      x3f_printf(DEBUG, "Calculate black level for %s\n", name[i]);
+    else
+      x3f_printf(DEBUG, "Do not calculate black level for %s\n", name[i]);
+
+  pixels_sum = 0;
+  black = alloca(colors*sizeof(uint64_t));
   black_sum = alloca(colors*sizeof(uint64_t));
-  memset(black_sum, 0, colors*sizeof(uint64_t));
-  pixels += sum_area(x3f, "DarkShieldTop", image, rescale, colors,
-		     black_sum);
-  if (use_bottom)
-    pixels += sum_area(x3f, "DarkShieldBottom", image, rescale, colors,
-		       black_sum);
-  if (pixels == 0) return 0;
+  for (i=0; i<colors; i++) black_sum[i] = 0;
 
-  for (i=0; i<colors; i++)
-    black_level[i] = (double)black_sum[i]/pixels;
-
-  pixels = 0;
-  black_sum_sqdev = alloca(colors*sizeof(double));
-  memset(black_sum_sqdev, 0, colors*sizeof(double));
-  pixels += sum_area_sqdev(x3f, "DarkShieldTop", image, rescale, colors,
-			   black_level, black_sum_sqdev);
-  if (use_bottom)
-    pixels += sum_area_sqdev(x3f, "DarkShieldBottom", image, rescale, colors,
-			     black_level, black_sum_sqdev);
-  if (pixels == 0) return 0;
+  x3f_printf(DEBUG, "Dark level\n");
+
+  for (i=0; i<4; i++)
+    if (use[i]) {
+      int color;
+      int pixels = sum_area(area[i], colors, black);
+
+      pixels_sum += pixels;
+
+      x3f_printf(DEBUG, "  %s (%d)\n", name[i], pixels);
+
+      for (color = 0; color < colors; color++) {
+	x3f_printf(DEBUG, "    mean[%d] = %f\n",
+		   color,
+		   (double)black[color]/pixels);
+	black_sum[color] += black[color];
+      }
+    }
+
+  if (pixels_sum == 0) return 0;
 
   for (i=0; i<colors; i++)
-    black_dev[i] = sqrt(black_sum_sqdev[i]/pixels);
+    black_level[i] = (double)black_sum[i]/pixels_sum;
+
+  pixels_sum = 0;
+  black_sqdev = alloca(colors*sizeof(double));
+  black_sqdev_sum = alloca(colors*sizeof(double));
+  for (i=0; i<colors; i++) black_sqdev_sum[i] = 0.0;
+
+  for (i=0; i<4; i++)
+    if (use[i]) {
+      int color;
+      int pixels = sum_area_sqdev(area[i], colors, black_level, black_sqdev);
+
+      pixels_sum += pixels;
+
+      x3f_printf(DEBUG, "  %s (%d)\n", name[i], pixels);
+
+      for (color = 0; color < colors; color++) {
+	x3f_printf(DEBUG, "    dev[%d] = %g\n",
+		   color,
+		   sqrt(black_sqdev[color]/pixels));
```

Comment:
As mentioned, this printout is incorrect, since it uses the overall mean to calculate the standard deviation for individual areas. This need to ether be fixed or removed. In the current state it more confusing than useful. To fix it, you would av do do an additional call too sum_area_squdev in the first loop that calculates the means. The result of this call would be used only for printouts.